File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/ScaleSymbol.java
package com.valentingrigorean.arcgis_maps_flutter.map;

import com.esri.arcgisruntime.symbology.CompositeSymbol;
import com.esri.arcgisruntime.symbology.PictureFillSymbol;
import com.esri.arcgisruntime.symbology.PictureMarkerSymbol;
import com.esri.arcgisruntime.symbology.Symbol;

import java.util.ArrayList;

public class ScaleSymbol extends Symbol {
    private final Symbol symbol;
    private final ArrayList<SymbolScaleHandle> scaleHandles = new ArrayList<>();

    public ScaleSymbol(Symbol symbol) {
        super(symbol.getInternal());
        this.symbol = symbol;
        populateDefaultSize(symbol);
    }


    public Symbol getSymbol() {
        return symbol;
    }

    public void setScale(float scale) {
        for (final SymbolScaleHandle scaleHandle : scaleHandles) {
            scaleHandle.setScale(scale);
        }
    }

    @Override
    protected void finalize() throws Throwable {
        super.finalize();
        scaleHandles.clear();
    }

    private void populateDefaultSize(Symbol symbol) {
        if (symbol instanceof CompositeSymbol) {
            CompositeSymbol compositeSymbol = (CompositeSymbol) symbol;
            if (compositeSymbol != null) {
                for (final Symbol child : compositeSymbol.getSymbols()) {
                    populateDefaultSize(child);
                }
            }
        } else {
            scaleHandles.add(new SymbolScaleHandle(symbol));
        }
    }

    private class SymbolScaleHandle {
        private final float width;
        private final float height;
        private final boolean haveSize;
        private final Symbol symbol;

        private float scale = 1f;

        public SymbolScaleHandle(Symbol symbol) {
            this.symbol = symbol;
            if (symbol instanceof PictureMarkerSymbol) {
                PictureMarkerSymbol pictureMarkerSymbol = (PictureMarkerSymbol) symbol;
                haveSize = true;
                width = pictureMarkerSymbol.getWidth();
                height = pictureMarkerSymbol.getHeight();
            } else if (symbol instanceof PictureFillSymbol) {
                PictureFillSymbol pictureFillSymbol = (PictureFillSymbol) symbol;
                haveSize = true;
                width = pictureFillSymbol.getWidth();
                height = pictureFillSymbol.getHeight();
            } else {
                haveSize = false;
                width = 0;
                height = 0;
            }
        }

        public void setScale(float scale) {
            if (!haveSize) {
                return;
            }
            if (this.scale == scale) {
                return;
            }
            this.scale = scale;
            final float newWidth = width * scale;
            final float newHeight = height * scale;
            if (symbol instanceof PictureMarkerSymbol) {
                PictureMarkerSymbol pictureMarkerSymbol = (PictureMarkerSymbol) symbol;
                pictureMarkerSymbol.setWidth(newWidth);
                pictureMarkerSymbol.setHeight(newHeight);
            } else if (symbol instanceof PictureFillSymbol) {
                PictureFillSymbol pictureFillSymbol = (PictureFillSymbol) symbol;
                pictureFillSymbol.setWidth(newWidth);
                pictureFillSymbol.setHeight(newHeight);
            }
        }
    }
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/ScaleSymbol.kt
package com.valentingrigorean.arcgis_maps_flutter.map

import com.arcgismaps.mapping.symbology.Symbol


class ScaleSymbol(val symbol: Symbol) : Symbol(
    symbol.internal
) {
    private val scaleHandles = ArrayList<SymbolScaleHandle>()

    init {
        populateDefaultSize(symbol)
    }

    fun setScale(scale: Float) {
        for (scaleHandle in scaleHandles) {
            scaleHandle.setScale(scale)
        }
    }

    @Throws(Throwable::class)
    protected fun finalize() {
        super.finalize()
        scaleHandles.clear()
    }

    private fun populateDefaultSize(symbol: Symbol) {
        if (symbol is CompositeSymbol) {
            val compositeSymbol = symbol
            if (compositeSymbol != null) {
                for (child in compositeSymbol.symbols) {
                    populateDefaultSize(child)
                }
            }
        } else {
            scaleHandles.add(SymbolScaleHandle(symbol))
        }
    }

    private inner class SymbolScaleHandle(private val symbol: Symbol) {
        private var width = 0f
        private var height = 0f
        private var haveSize = false
        private var scale = 1f

        init {
            if (symbol is PictureMarkerSymbol) {
                val pictureMarkerSymbol = symbol
                haveSize = true
                width = pictureMarkerSymbol.width
                height = pictureMarkerSymbol.height
            } else if (symbol is PictureFillSymbol) {
                val pictureFillSymbol = symbol
                haveSize = true
                width = pictureFillSymbol.width
                height = pictureFillSymbol.height
            } else {
                haveSize = false
                width = 0f
                height = 0f
            }
        }

        fun setScale(scale: Float) {
            if (!haveSize) {
                return
            }
            if (this.scale == scale) {
                return
            }
            this.scale = scale
            val newWidth = width * scale
            val newHeight = height * scale
            if (symbol is PictureMarkerSymbol) {
                val pictureMarkerSymbol = symbol
                pictureMarkerSymbol.width = newWidth
                pictureMarkerSymbol.height = newHeight
            } else if (symbol is PictureFillSymbol) {
                val pictureFillSymbol = symbol
                pictureFillSymbol.width = newWidth
                pictureFillSymbol.height = newHeight
            }
        }
    }
}